### PR TITLE
[NearestNeighbors] Avoid changing the signature of "_fit" and move kneighbor API to the model class

### PR DESCRIPTION
**Pull request by lijinf2**

### CODE REVIEW

File: src/spark_rapids_ml/core.py
Comment (eordentlich) [inline src/spark_rapids_ml/core.py:480]: This is inherited from _CumlCaller now so don't think it is needed here again.

File: src/spark_rapids_ml/core.py
Comment (eordentlich) [inline src/spark_rapids_ml/core.py:545]: Why order change here?

**Review by eordentlich — COMMENTED** [review]
Looks good.  Couple minor comments.  Make sure to run tests.

File: src/spark_rapids_ml/core.py
Comment (leewyang) [inline src/spark_rapids_ml/core.py:241]: Please add some pydoc to explain why this class exists (for future developers).

File: src/spark_rapids_ml/knn.py
Comment (leewyang) [inline src/spark_rapids_ml/knn.py:87]: "a row"

File: src/spark_rapids_ml/knn.py
Comment (leewyang) [inline src/spark_rapids_ml/knn.py:228]: We should test this at scale to see if/when we hit limits.

File: src/spark_rapids_ml/knn.py
Comment (leewyang) [inline src/spark_rapids_ml/knn.py:266]: do we still need `# type: ignore` here?

File: src/spark_rapids_ml/knn.py
Comment (leewyang) [inline src/spark_rapids_ml/knn.py:314]: Should add pydocs for this method since it's a new method, esp. explaining arguments and return values.

File: tests/test_nearest_neighbors.py
Comment (leewyang) [inline tests/test_nearest_neighbors.py:117]: should add a test that asserts that an exception is raised if `transform` is invoked.

File: src/spark_rapids_ml/core.py
Comment (lijinf2) [reply src/spark_rapids_ml/core.py]: Yeah. Confirmed it is not needed. Already have it deleted.

File: src/spark_rapids_ml/core.py
Comment (lijinf2) [reply src/spark_rapids_ml/core.py:544]: Long story short, this relates to python MRO (Method Resolution Order). 

Specifically, order change here to keep the same inheritance order as in _CumlCaller(_CumlParams, _CumlCommon). Otherwise, a python MRO  error will appear. 

We must use inheritance order _CumlCaller(_CumlParams, _CumlCommon) because reverse order _CumlCaller(_CumlCommon, _CumlParams) triggers a MRO error in _CumlEstimator(Estimator, _CumlCaller). The error message is "TypeError: Cannot create a consistent method resolution order (MRO) for bases Params, Generic".  It seems because in _CumlCaller superclass Generic appears before superclass Params. But in Estimator superclass Params appears before the superclass Generic.

File: src/spark_rapids_ml/core.py
Comment (lijinf2) [reply src/spark_rapids_ml/core.py:241]: Added some explanations.

File: src/spark_rapids_ml/knn.py
Comment (lijinf2) [reply src/spark_rapids_ml/knn.py]: good catch

File: src/spark_rapids_ml/knn.py
Comment (lijinf2) [reply src/spark_rapids_ml/knn.py:229]: Yeah, marked in Jira ticket DML-457

File: src/spark_rapids_ml/knn.py
Comment (lijinf2) [reply src/spark_rapids_ml/knn.py:267]: Yeah. This is because _CumlCaller has an abstractmethod _out_schema(self), and _CumlModel also has an abstractmethod _out_schema(self, input_schema: StructType).  

Was not able to implement both methods in NearestNeighborsModel, yet uncomment the # type: ignore will trigger a mypy error.

File: src/spark_rapids_ml/knn.py
Comment (lijinf2) [reply src/spark_rapids_ml/knn.py:315]: done. Thanks for the advice!

File: tests/test_nearest_neighbors.py
Comment (lijinf2) [reply tests/test_nearest_neighbors.py:120]: Added the following:

```python
    def _transform(self, dataset: DataFrame) -> DataFrame:
        raise NotImplementedError(
            "'NearestNeighborsModel does not provide a transform function. Use 'kneighbors' instead."
        )
```

Tested in pyspark shell that the error will be raised if transform is invoked. 

Did not find the solution on how to add assert statement for exception. Let me know.

**Comment by lijinf2** [comment]
Revised the PR. 
The revised version got run_test.sh tested and all tests passed. Also got "pytest -ra --runslow --durations=10 tests" tested and all tests passed.

File: src/spark_rapids_ml/core.py
Comment (wbo4958) [inline src/spark_rapids_ml/core.py:360]: is `return_model` needed here?

File: src/spark_rapids_ml/core.py
Comment (wbo4958) [inline src/spark_rapids_ml/core.py:241]: maybe _CumlFitCaller?

File: src/spark_rapids_ml/knn.py
Comment (wbo4958) [inline src/spark_rapids_ml/knn.py:265]: if the variables of a class are not willing to be exposed, please add "_", eg, self._item_df ...

File: src/spark_rapids_ml/knn.py
Comment (wbo4958) [inline src/spark_rapids_ml/knn.py:305]: Please make sure if the label type should be converted to be the same as features type.

File: src/spark_rapids_ml/knn.py
Comment (wbo4958) [inline src/spark_rapids_ml/knn.py:316]: long length for a single line. Could you rework this comment?

File: src/spark_rapids_ml/knn.py
Comment (wbo4958) [inline src/spark_rapids_ml/knn.py:393]: query_filter = label_array == 1 ?

File: src/spark_rapids_ml/knn.py
Comment (wbo4958) [inline src/spark_rapids_ml/knn.py:497]: Is this error message for the end user?

File: tests/test_nearest_neighbors.py
Comment (leewyang) [reply tests/test_nearest_neighbors.py:120]: See [this example](https://github.com/NVIDIA/spark-rapids-ml/blob/spark-cuml/tests/test_kmeans.py#L317-L318).

File: src/spark_rapids_ml/knn.py
Comment (leewyang) [inline src/spark_rapids_ml/knn.py:487]: Extra single quote `'NearestNeighborsModel`, or needs trailing quote, `'NearestNeighborsModel'`

File: tests/test_nearest_neighbors.py
Comment (lijinf2) [reply tests/test_nearest_neighbors.py:121]: Done.

File: src/spark_rapids_ml/core.py
Comment (lijinf2) [reply src/spark_rapids_ml/core.py]: done

File: src/spark_rapids_ml/core.py
Comment (lijinf2) [reply src/spark_rapids_ml/core.py]: done

File: src/spark_rapids_ml/core.py
Comment (lijinf2) [reply src/spark_rapids_ml/core.py:241]: Yeah, mostly sparkcuml calls fit functions, but it also calls kneighbor that belongs to NearestNeighborModel class.  Maybe we keep _CumlCaller for now and decide if we should rename in the future.

File: src/spark_rapids_ml/knn.py
Comment (lijinf2) [reply src/spark_rapids_ml/knn.py]: done

File: src/spark_rapids_ml/knn.py
Comment (lijinf2) [reply src/spark_rapids_ml/knn.py:305]: Yeah, it seems the current type is sufficient to differentiate query vectors from item vectors.

File: src/spark_rapids_ml/knn.py
Comment (lijinf2) [reply src/spark_rapids_ml/knn.py]: done.

File: src/spark_rapids_ml/knn.py
Comment (lijinf2) [reply src/spark_rapids_ml/knn.py]: done. Good suggestion.

File: src/spark_rapids_ml/knn.py
Comment (lijinf2) [reply src/spark_rapids_ml/knn.py:498]: Yeah. Currently, the Model class of KNN does not support the transform function.

File: src/spark_rapids_ml/knn.py
Comment (lijinf2) [reply src/spark_rapids_ml/knn.py]: Good catch. Done.

**Comment by lijinf2** [comment]
Finished the 2nd round revision on the PR. 
Got run_test.sh tested and all tests passed.

File: src/spark_rapids_ml/core.py
Comment (eordentlich) [inline src/spark_rapids_ml/core.py:327]: communicaiton -> communication

File: src/spark_rapids_ml/knn.py
Comment (eordentlich) [inline src/spark_rapids_ml/knn.py:326]: 'that each' -> 'where each'

File: src/spark_rapids_ml/knn.py
Comment (eordentlich) [inline src/spark_rapids_ml/knn.py:343]: of knn -> of knn item_df vectors.

**Comment by lijinf2** [comment]
@eordentlich Thanks! I have just fixed the typo and the wording.

**Review by eordentlich — APPROVED** [review]
Make sure run_tests.sh still passes before merging.

👍

**Comment by lijinf2** [comment]
> Make sure run_tests.sh still passes before merging.
> 
> 👍

Yeah. run_tests.sh had been passed. I just rebased and run_tests.sh got passed again with no problem. Need approval again because of the rebase.

**Review by eordentlich — APPROVED** [review]
👍

**Pull request merged**
